src/parser/core/Module.ts

```diff
@@ -0,0 +1,273 @@
+		throw new Error(`${constructor.name}'s dependency \`${prop}\` is invalid. Expected \`Module\`, got \`${dependency.name}\`.`)
+	}
+
+	constructor.dependencies.push({
+		handle: dependency.handle,
+		prop,
+	})
+}
+
+interface MappedDependency {
+	handle: string
+	prop: string
+}
+
+type Filter<T extends Event> = Partial<T> & Partial<{
+	abilityId: Ability['guid']
+	to: 'player' | 'pet' | T['targetID']
+	by: 'player' | 'pet' | T['sourceID']
+}>
+
+type HookCallback<T extends Event> = (event: T) => void
+
+interface Hook<T extends Event> {
+	events: T['type'][]
+	filter: Filter<T>
+	callback: HookCallback<T>
+}
+
+export default class Module {
+	static dependencies: Array<string | MappedDependency> = []
+	static displayOrder = DISPLAY_ORDER.DEFAULT
+
+	private static _handle: string
+	static get handle() {
+		if (!this._handle) {
+			const handle = this.name.charAt(0).toLowerCase() + this.name.slice(1)
+			console.error(`\`${this.name}\` does not have an explicitly set handle. Using \`${handle}\`. This WILL break in minified builds.`)
+			this._handle = handle
+		}
+		return this._handle
+	}
+	static set handle(value) {
+		this._handle = value
+	}
+
+	private static _title: string
+	static get title() {
+		if (!this._title) {
+			this._title = this.handle.charAt(0).toUpperCase() + this.handle.slice(1)
+		}
+		return this._title
+	}
+	static set title(value) {
+		this._title = value
+	}
+
+	// DI FunTimes™
+	[key: string]: any;
+
+	private _hooks = new Map<Event['type'], Set<Hook<Event>>>()
+
+	constructor(
+		protected readonly parser: Parser
+	) {
+		const module = this.constructor as typeof Module
+		module.dependencies.forEach(dep => {
+			if (typeof dep === 'string') {
+				dep = {handle: dep, prop: dep}
+			}
+
+			this[dep.prop] = parser.modules[dep.handle]
+		})
+		this.init()
+	}
+
+	// So TS peeps don't need to pass the parser down
+	protected init() {}
+
+	normalise(events: Event[]) {
+		return events
+	}
+
+	/**
+	 * This method is called when an error occurs, either when running
+	 * event hooks or calling {@link Module#output} in this module or
+	 * a module that depends on this module.
+	 * @param {String} source Either 'event' or 'output'
+	 * @param {Error} error The error that occurred
+	 * @param {Object} event The event that was being processed when the error occurred, if source is 'event'
+	 * @returns {Object|undefined} The data to attach to automatic error reports, or undefined to rely on primitive value detection
+	 */
+	getErrorContext(source: 'event' | 'output', error: Error, event?: Event): any {
+		return
+	}
+
+	protected addHook<T extends Event>(
+		events: T['type'] | T['type'][],
+		cb: HookCallback<T>
+	): Hook<T>
+	protected addHook<T extends Event>(
+		events: T['type'] | T['type'][],
+		filter: Filter<T> | HookCallback<T>,
+		cb: HookCallback<T>,
+	): Hook<T>
+	protected addHook<T extends Event>(
+		events: T['type'] | T['type'][],
+		_filter: Filter<T> | HookCallback<T>,
+		_cb?: HookCallback<T>,
```

Comment:
I generally prefer to name them with an "Arg" suffix (`filterArg`, `cbArg`) because the `_` prefix is otherwise used to mark _unused_ arguments.